### DISCUSSION TITLE
[Feature Request]: Handle different TLDs better

### DISCUSSION THREAD

**Discussion by Gelbotron** [beyondcode/herd-community#376] [Ideas]
### Feature Description

As a developer using multiple devices for development, I encountered an issue after installing Herd. It took me considerable time to identify the source of the problem - a TLD conflict. Herd displayed the `.test` domain in the Sites tab, which confused me, as my setup on the personal computer used the `.localhost` TLD.

While I understand that supporting random TLDs might not be feasible, I suggest implementing a warning system within Herd to alert users if their TLD setup, specifically `.test`, is not configured correctly.

### Is this feature valuable for other users as well and why?

This improvement would help avoid confusion and streamline the development process for Herd users.

**Comment by sujithtomy** [comment]
a ".test" tld will not accept in Google Auth Client. so it is a deadlock situation to develop laravel google auth projects with Herd as far as I know. (any other solutions?) 

> Not sure why this ticket is closed with out a fix or suggestion to workaround!

**Reply by matthewhutchings** [reply]
having this issue - google auth requires a tld or localhost. 

I cant seem to access my project via localhost (when set in env) but can access it through .test

The only way round this was to use php artisan share but its a lot slower.

**Reply by maximepvrt** [reply]
Feel free to upvote and leave a comment on Issue https://issuetracker.google.com/issues/313466544?pli=1 to push Google to allow .test domains in OAuth callback URLs 🚀

**Reply by sschlein** [reply]
We've implemented a way to solve that for Google auth but still keep .test domains in Herd: https://herd.laravel.com/docs/windows/advanced-usage/social-auth

**Comment by milhouse1337** [comment]
I used the `herd tld` command to set a custom domain that resolves to `127.0.0.1`. But it doesn't seems to be "officially" supported. 🤷‍♂️

```
➜  ~ herd tld example.com
Using a custom TLD is no longer officially supported and may lead to unexpected behavior. Do you wish to proceed? [y/N]
```

Maybe there is a better way, but so far it worked like a charm for me.

**Comment by maximepvrt** [comment]
@mechelon do you plan to add this feature soon? same problem as @sujithtomy, ".test" tld will not accept in Google Auth Client

**Comment by 199ocero** [comment]
Is there any update for this? I want also to use Herd but it's preventing me to try since I have a Google auth and `.test` domain is not working.

**Reply by maximepvrt** [reply]
Feel free to upvote and leave a comment on Issue https://issuetracker.google.com/issues/313466544?pli=1 to push Google to allow .test domains in OAuth callback URLs 🚀

**Comment by JonPurvis** [comment]
I'd also like to see this feature implemented. Having to run some projects locally outside of Herd and it is not fun 😭

**Comment by michaloravec** [comment]
This is a necessary feature.

**Comment by joshlewis** [comment]
Agreed, this needs to be easier.

**Comment by maximepvrt** [comment]
@mechelon @sschlein @mpociot This was announced in May, but still not available 😢https://x.com/marcelpociot/status/1790412281307701466

**Reply by maximepvrt** [reply]
@mechelon @sschlein @mpociot Why is there still no news about this feature?

**Comment by HichemTab-tech** [comment]
any progress here ? I'm blocked in the same situation, can't use `.test` domain in google auth api. i can just "**not use**" Herd to resolve my problem, but i really do want to use it for other feature, please don't ruin this just because of this issue.

**Reply by maximepvrt** [reply]
Feel free to upvote and leave a comment on Issue https://issuetracker.google.com/issues/313466544?pli=1 to push Google to allow .test domains in OAuth callback URLs 🚀

**Comment by jonappcinking** [comment]
What steps can we take to address these issues? As a developer, I find working with the Google Console to be a bottleneck.

**Reply by sschlein** [reply]
We're currently testing this feature internally but it's pretty complex as a change of the tld needs regeneration of all config files and certificates. Should arrive soon

**Comment by rifrocket** [comment]
I wan to contribute to solving this issue, can I ?, because I need the functionality to update TLD, it is very important feature.

**Reply by sschlein** [reply]
You can use Expose to create a shared url until this feature is ready and even get a .com that points to your local app

**Comment by ejunker** [comment]
I used the `herd tld` to set a custom TLD and so far it seems to be working well. I hope that in the future the custom TLD will be able to be configured in the `herd.yml` file.

**Comment by sschlein** [comment]
The problem with custom tlds (right now) is that they also effect TLS certificate generations and all existing nginx configurations, so it's not the easiest feature and just running `herd tld` could break your setup at the moment.

The feature is planned on a global basis and we hope to release it with Herd 1.12 – but this also means that individual tld configuration per site like it's done in the `herd.yml` wouldn't work.

**Reply by maximepvrt** [reply]
any news about this?

**Comment by nshontz** [comment]
Bumping this issue. 

As mentioned before, a ".test" tld is not accepted in Google Auth Client, making apps that use Google for Auth incompatible with Herd. 

Thanks for all of the work. Herd is amazing.

**Reply by maximepvrt** [reply]
Feel free to upvote and leave a comment on Issue https://issuetracker.google.com/issues/313466544?pli=1 to push Google to allow .test domains in OAuth callback URLs 🚀

**Comment by badrshs** [comment]
Has anyone come across a solution for this? The` herd tld dev`  didn't work for me. 

and yeh same issue as the guys mentioned Google auth doesn't allow .test domain , I'm trying to use .dev or anything else

**Reply by maximepvrt** [reply]
Feel free to upvote and leave a comment on Issue https://issuetracker.google.com/issues/313466544?pli=1 to push Google to allow .test domains in OAuth callback URLs 🚀

**Reply by jakebathman** [reply]
Please don't spam the same comment on every reply, people subscribed to this issue get an email for every single one.

**Comment by sergyu91ma** [comment]
What about not using a TLD at all? This should be optional. 
I have apps with `"local<sitename>"` for local development and i really want to leave them like that when i switch to herd. 
Why should be a .smth mandatory ? 
Maybe i don't use ssl locally, if that is a problem. 
Anyway this should be at least configurable if not optional. 
In other tools like xampp, laragon, flydev this is not a problem. You can customize it however you want.

**Comment by HichemTab-tech** [comment]
guys, the solution proposed by @milhouse1337 works pretty well; you shouldn't be blocked by this error :) I waited too long for a solution either from both the Laravel team and Google, but when I tried this solution I stopped worrying about whether they fix it or not haha.

**Reply by badrshs** [reply]
it doesn't work,  at least not for everyone , otherwise it will be officially documented

**Reply by sergyu91ma** [reply]
![image](https://github.com/user-attachments/assets/7c95a660-bc6b-4a11-aff8-407c11bdefde)

After changing TLD in hosts files it still generated .test even though in UI says .local, so it doesn't work. 

But what about if i don't want TLD at all? how to disable it? is it possibe?

**Reply by HichemTab-tech** [reply]
Yes exactly Herd is still adding .test to host file, so u need to add localanvelope.local manually to the host file, and it should be outside the `# generated by Herd` section.

And it should work !

Like we are tricking Herd to link the .local domain with app instead of .test, but it is still adding .test to host, not .local. that's why we need to add it manually.

Hope i explained well.

**Reply by HichemTab-tech** [reply]
Although, about not using TLD at all, I don't really know cuz i didn't really tried it, but the way I'm imagining it is using external tool like xampp :)

**Reply by sergyu91ma** [reply]
It partially works like you say, because if i want to change the php version for a site that won't work. 
It will get the global version of php used by herd.  I just tested that. 

Ex: if i want for localanvelope.local php 8.4, but globally i have 8.1, accessing localanvelope.local in browser will be done with php 8.1 since herd knows that for localanvelope.test should use php 8.4 
If you want SSL, i think will have the same problem.

**Reply by HichemTab-tech** [reply]
Wait, by php 8.1 you mean the php installed outside Herd ?

**Reply by sergyu91ma** [reply]
no. php 8.1 and 8.4 installed from herd. see picture attached to understand. 
You can have multiple php version installed and use a specific version  for a specific site

![image](https://github.com/user-attachments/assets/38416e7e-7429-412c-adf1-6ff00491463f)

**Reply by HichemTab-tech** [reply]
Ah okay i understand now, well actually that's a good observation, i didn't noticed that before, I'll check it. So technically we didn't fully trick Herd didn't we, lol.

**Reply by sergyu91ma** [reply]
I don't think we can fully trick it. It has to be implemented to have a custom TLD or to disable it, not having at all.

**Reply by badrshs** [reply]
Ok that worked now :D , I tried your approach of changing the host file and it worked for me finally, thanks 
so basically what I did was the following 

 herd tld me
 
 then I went to hosts and updated the domain there with .me and done

**Reply by sergyu91ma** [reply]
Works until herd restart. It will regenerate the sites in the hosts after restart.

**Reply by HichemTab-tech** [reply]
No actually, the trick here is that u keep your host outside the section where it says `# generated by Herd`.

This way it stays there even if you restart Herd.

**Reply by HichemTab-tech** [reply]
It's `# Herd generated Hosts. Do not change`. Actually.

So for example put your `example.com` before this line.

**Reply by badrshs** [reply]
Yep glade you mentioned this, worked fine after the restart

![image](https://github.com/user-attachments/assets/4026c264-5922-4612-98e1-6c529b2d6880)

**Reply by sergyu91ma** [reply]
Yes, but this works until you have to change php version for a site to be different than global, as explained above with images. 
Then if global doesn't have xdebug, but the isolated has you cannot use xdebug either. Ssl same problem. This is a partial workaround that works in simple cases.

**Reply by sschlein** [reply]
Hey, while you're trying to trick Herd, you seem to actually get an idea why this hasn't been released yet 😅

As you realize, it's pretty complex to make this work in a reliable and good way, not even mentioning that this update would have to update nginx and dnsmasq (on macOS) configurations for 40,000+ developers, so releasing a version that breaks how Herd actually resolves sites would be a nightmare and kill productivity for so many people.

This is why we don't rush this but will eventually come up with a solution that works great for you.

We're in the process of a bigger update for other features and Herd Pro services but think that custom TLDs have a chance to be on the roadmap soon!

**Reply by sergyu91ma** [reply]
I get that is complex changing/having user defined TLDs, but why there is no option to be disabled? 
Without TLDs things are even more simpler, you don't have this problems at all. 
Not everyone wants TLDs locally, not to say that they have to use .test mandatory, like is a hardcoded thing everywhere in the app.  
Keeping things simpler is better for everyone.

**Reply by sschlein** [reply]
No TLD is not on the roadmap as this breaks the whole idea behind Herd running on actual domains, so I don't think we will reconsider this in the future. If this is a requirement for you, Herd might not be a good fit for you.

You are actually the first one requesting this.

**Reply by HichemTab-tech** [reply]
@sergyu91ma I just tried what you told me yesterday about the php versions, and it worked, I didn't face a problem :) i mean i set php 8.3 for my site, and the global version is 8.4, but the site IS actually running on 8.3 as expected i mean Herd worked perfectly, look : 

![image](https://github.com/user-attachments/assets/329f1006-c21a-456f-b57a-d8b6e2b2faf5)

![image](https://github.com/user-attachments/assets/362ecf4d-a3ce-4136-8ff7-df12685d57ef)


- my route to show php info : 

![image](https://github.com/user-attachments/assets/a6baec17-2757-46ae-b9ec-163efea6a254)

- the output :D 

![image](https://github.com/user-attachments/assets/65d8ffbf-8529-4993-b28c-f88083c0d6da)



so the site followed the right version even if global version is different :D

**Reply by sergyu91ma** [reply]
Looks like your TLD is .com in herd, different from default, and weirdly is working. 
Then smth is different on my side. 

I did use link in herd for the sites. Maybe because of that?! 
I didn't used parked path. From pictures you use that. 

Can you also try to use xdebug on the 8.3 version? to see if it works?

**Reply by HichemTab-tech** [reply]
> I did use link in herd for the sites. Maybe because of that?! I didn't used parked path. From pictures you use that.
 what do you mean by link ?
 And i just tried xdebug and it worked :")

**Reply by sergyu91ma** [reply]
You can link a site (which makes a symlink behind) if you want another domain name for your site rather than the folder name. 
Read here in docs in the second part.
https://herd.laravel.com/docs/windows/getting-started/sites 

I did it like that and it didnt't got the right php version when isolated.

**Comment by RonDarby** [comment]
I'm looking for a solution that will allow me to have multiple tlds concurrently in herd, where I can select which one I want for each project. Not here to bump the discussion, but reading the issues everyone here is mentioning about google auth, I managed to solve that by simply doing the following:

- `cd ~/Library/Application\ Support/Herd/config` you will find the various services,  nginx, dnsmasq, valet and phpfpm files etc
- `vim ./valet/config.json` and edit the `tld` value to `localtest.me`
- cd into your project folder and `herd link what-have-you` or `herd init` and follow the prompts
- https://what-have-you.localtest.me works
- DO NOT go `herd tld xxxx` unless you are happy to break any custom setups you may have, I had a few custom nginx files serving as a proxy to nodejs servers running in docker projects that needed fixing up again (I like the self signed Valet CA certificates for local testing)

Sorted, no edit of hosts files or anything as xxxx.localtest.me points to 127.0.0.1 and google recognises it.

I tried editing dnsmasq with no joy, I'll still crack that one, but the certs will still be a problem with other domains that weren't added via the `herd link` command, there are limitations though.

**Comment by mikebronner** [comment]
Additionally, VSCode no longer follows links with `.test` domains in the terminal, as they are not valid URLs. VSCode now validates the URLs apparently.

This is definitely an annoyance, and manually setting the TLD via the command outlined in previous commends `herd tld` doesn't seem to work anymore, as the sites show as your custom TLD, but still resolve to `.test`.

**Comment by sschlein** [comment]
We've built a small service to make implementing Google auth (and other) easy: 
https://herd.laravel.com/docs/windows/advanced-usage/social-auth

I am pretty sure that VSCode will fix that soon because using a valid public TLD would introduce a serious security issue and while we have a prototype that supports other TLDs like `.wip`, `.herd` and `.local`, that doesn't solve the valid URL problem.

**Reply by mikebronner** [reply]
~Thanks for the explanation @sschlein, just wanted to add that `.local` domains do work with VSCode, we use those with our non-Herd projects.~ (This probably was because the domain was trusted.)

**Reply by mikebronner** [reply]
One more quick note in case others are experiencing the same issue I have: https://github.com/microsoft/vscode/issues/249353#issuecomment-2906690183

Adding your `.test` domain to VSCode's Trusted Domains seems to solve the problem and make them clickable in the terminal. Or adding `"*",`  to the top of the list seems to work as well.

**Reply by pktharindu** [reply]
@sschlein, Just out of curiosity, what's the actual security issue with using public TLDs for local development?

**Comment by sschlein** [comment]
Imagine using .com domains for development and overwriting them in your hosts file automatically – an attacker could somehow try to use Herds tooling and redirect your banks domain, paypal, etc to their site and you wouldn't even notice.

Even if the chances may be small, we just want this to be impossible.

**Reply by pktharindu** [reply]
Ah interesting, thanks for explaining! Though I'm curious - if an attacker has enough access to abuse Herd's tooling, couldn't they just edit `/etc/hosts` directly anyway? Is this more of a defense-in-depth thing to limit the blast radius if something goes wrong?

**Reply by sschlein** [reply]
That plays into it, I just don't want to be even remotely legally responsible for providing an attack vector on tens of thousands of machines 😅

There is also very limited need for the real TLD as local dev and we provide ways around this for the cases where 3rd party services require specific TLDs.

**Reply by pktharindu** [reply]
That's fair! Again, thanks for taking the time!

**Reply by nie7321** [reply]
> There is also very limited need for the real TLD as local dev and we provide ways around this for the cases where 3rd party services require specific TLDs.

I still have several legacy apps that require a cookie set for our base domain. 

Is support for `herd tld` still planned?
